### ISSUE TITLE
undefined field values with typescript 4.3.x

### ISSUE THREAD

**Issue by nowrap** [vuex-orm/vuex-orm#734]
#### Describe the bug

Using typescript 4.3.x results in undefined field values after creating, inserting or dispatching data to a model

#### Steps to reproduce the bug

1. Checkout the TS example https://github.com/vuex-orm/vuex-orm-examples-ts
2. update the typescript from 3.9.x to latest
3. npm run serve
4. See undefined fields in both lists

#### Expected behaviour

See the created or inserted data in both lists

#### Versions

- Vuex ORM: ^0.36.4
- Vue: ^2.6.12
- Vuex: ^3.6.2

#### Link to minimal reproduction

https://github.com/vuex-orm/vuex-orm-examples-ts

#### Additional context

Downgrading to typescript 4.2.x solves this issue

**Comment by cuebit** [comment]
This worked up until, and including, `4.3.0-beta`. It seems [`2484210`](https://github.com/microsoft/TypeScript/commit/2484210a00774e8f58789ad3f19e66affaa44de2) changed the default [`useDefineForClassFields`](https://www.typescriptlang.org/tsconfig#useDefineForClassFields) for `ESNext` targets. So this only impacts `ESNext`, sadly. I can't tell for sure if this is intended behaviour, 

As a workaround, either:
1. set `useDefineForClassFields` to `true` if using `esnext`
2. downgrade to 4.2
3. set a target of `ES2018` at least (unless you really need `ES5` 👀 )

**Comment by cuebit** [comment]
This worked up until, and including, `4.3.0-beta`. It seems [`2484210`](https://github.com/microsoft/TypeScript/commit/2484210a00774e8f58789ad3f19e66affaa44de2) changed the default [`useDefineForClassFields`](https://www.typescriptlang.org/tsconfig#useDefineForClassFields) for `ESNext` targets. So this only impacts `ESNext`, sadly. I can't tell for sure if this is intended behavior or a bug. Looks like scope-creep to me as this was in the 4.3.1 milestone but moved to 4.4 🤷‍♂️ 

As a workaround, either:
1. set `useDefineForClassFields` to `true` if using `esnext` (or `false` as this seems to work for vite users)
2. downgrade to 4.2
3. set a target of `ES2018` at least (unless you really need `ES5` 👀 )

**Comment by JonasJW** [comment]
I'm experiencing the same issue but none of @cuebit workarounds seem to work for me

**Comment by Papooch** [comment]
Same here. Neither setting different target nor the `useDefineForClassFields` flag to true fixes this :( Downgrading typescript is sadly not an option in my case.
@JonasJW have you managed to fix it it somehow?

**Comment by JonasJW** [comment]
> Same here. Neither Setting different target nor the `useDefineForClassFields` flag to true fixes this :( Downgrading typescript is sadly not an option in my case.
> @JonasJW have you managed to fix it it somehow?

Yes I fixed it somehow. Not sure what solved it in the end. I'm using `"typescript": "4.1.2"` and in tsconfig.json I still compile to "esnext" and do not specify  `useDefineForClassFields`. I have upgraded various packages, I think that is what solved it in the end

**Comment by cuebit** [comment]
> I'm using `"typescript": "4.1.2"`...

@JonasJW that is precisely why it works for you now. This issue is more evident in TS 4.3 👍 

@Papooch what typescript version specifically? Also, could you share your tsconfig?

**Comment by Papooch** [comment]
@cuebit Thank you for the prompt reaction, I'm using typescript `4.3.5` in a monorepo project (with nx), this is the base tsconfig:
```json
{
    "compilerOptions": {
        "rootDir": ".",
        "sourceMap": true,
        "declaration": false,
        "moduleResolution": "node",
        "emitDecoratorMetadata": true,
        "experimentalDecorators": true,
        "importHelpers": true,
        "esModuleInterop": true,
        "target": "es2015",
        "module": "esnext",
        "lib": [
            "es2017",
            "dom"
        ],
        "skipLibCheck": true,
        "skipDefaultLibCheck": true,
        "noImplicitThis": true,
       "pahts": {}
}
```
And this is the extended tsconfig int the app's folder:
```js
{
    "extends": "../../tsconfig.json",
    "compilerOptions": {
        "outDir": "../../dist/out-tsc",
        "types": ["webpack-env", "node"],
        "target": "ES6", // this does no fix the issue
        "module": "esnext",
        "strict": true,
        "jsx": "preserve",
        "moduleResolution": "node",
        "useDefineForClassFields": true // and neither does this
},
```

**Comment by JonasJW** [comment]
@cuebit @Papooch for anything above 4.2.x didn't seem to work

**Comment by Papooch** [comment]
@cuebit I understand there's no intention to add fixes and features to vuex-orm atm, but do you happen to know if the `Vuex ORM Next` fixes this issue?

As a workaroud, I'm using this monstrosity to have types on the returned model:

I have to extend the original Model to include the attributes, but I can't use type parameters for static methods, therefore I use a class factory like this:

```ts
// vuex-orm-typed-model-factory.ts
function TypedModelFactory<TAttrs>() {
    class TypedModel extends Model {
        static all<T extends typeof Model>(
            this: T
        ): Collection<InstanceOf<T> & TAttrs> { // notice InstanceOf<T> & TAttrs
            return super.all() as any;
        }

        static find<T extends typeof Model>(
            this: T, id: string | number | (number | string)[]
        ): Item<InstanceOf<T> & TAttrs> {
            return super.find(id) as any;
        }

       // ... and other finder methods that I need

    }
    return TypedModel
}
```
And I define the models like this:

```ts
//  user.model.ts

export interface IUserAttrs {
  id!: string
  name!: string
  todos!: Todo[]
}

export default class User extends TypedModelFactory<IUserAttrs>() {
  static entity = 'users'

  static fields () {
    return {
      id: this.uid(),
      name: this.string(''),
      todos: this.hasMany(Todo, 'user_id')
    }
  }
}
```

**Comment by NorthBlue333** [comment]
Hello,
I ran into the same issue while upgrading my project dependencies.

After reading TypeScript documentation about `useDefineForClassFields` ([tsconfig](https://www.typescriptlang.org/tsconfig#useDefineForClassFields) and [changelog](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#the-usedefineforclassfields-flag-and-the-declare-property-modifier)), I've found this :

> Declarations are initialized with Object.defineProperty.
> Declarations are always initialized to undefined, even if they have no initializer.

I think this is where all fields are set to undefined. The solution for me has been to set `useDefineForClassFields` in `tsconfig.json` `compilerOptions` to `false` (the default is `true` if target is ES2022 or higher, including ESNext as specified on the documentation @cuebit). **The default behavior changed since TypeScript `4.3` (https://www.typescriptlang.org/docs/handbook/release-notes/typescript-4-3.html#usedefineforclassfields-now-defaults-to-true-on-esnext-and-eventually-on-es2022)**.

This is an issue, we should not have to set this flag to false, because, as specified in the documentation (again) : 

> The latest version of the upcoming specification has a different runtime behavior to TypeScript’s implementation but the same syntax.

**Comment by nowrap** [comment]
I can confirm, that using typescript 4.5 and useDefineForClassFields false is working

**Comment by timo-de-winter** [comment]
> I can confirm, that using typescript 4.5 and useDefineForClassFields false is working

This worked for me :)
